desc/objectives/_coils.py

```diff
@@ -850,6 +858,258 @@ def body(k):
+        then supply ``None``.
+
+    """
+
+    __doc__ = __doc__.rstrip() + collect_docs(
+        target_default="``target=0``.",
+        bounds_default="``target=0``.",
+        coil=True,
+    )
+
+    _scalar = False
+    _units = "(T)"
+    _print_value_fmt = "Maximum field on coil: "
+
+    def __init__(
+        self,
+        coil,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        loss_function=None,
+        deriv_mode="auto",
+        field=None,
+        component="mag",
+        xsection_grid=None,
+        centerline_grid=None,
+        name="field on coil",
+        jac_chunk_size=None,
+        use_softmax=False,
+        softmax_alpha=1.0,
+        bs_chunk_size=None,
+    ):
+        from desc.coils import AbstractFiniteBuildCoil, CoilSet
+
+        if target is None and bounds is None:
+            target = 0
+
+        self._xsection_grid = xsection_grid
+        self._centerline_grid = centerline_grid
+        self._use_softmax = use_softmax
+        self._softmax_alpha = softmax_alpha
+        self._has_field = field is not None
+        self._field = [field] if not isinstance(field, list) else field
+        self._bs_chunk_size = bs_chunk_size
+
+        errorif(
+            not type(coil) is CoilSet,
+            ValueError,
+            "coil must be of type CoilSet, not an individual Coil. Mixed coil sets "
+            "are not supported.",
+        )
+        errorif(
+            not all([isinstance(c, AbstractFiniteBuildCoil) for c in coil.coils]),
+            ValueError,
+            "All coils in the CoilSet must inherit from AbstractFiniteBuildCoil",
+        )
+        errorif(
+            component not in {"mag", "t", "p", "q"},
+            ValueError,
+            "Component must be one of 'mag', 't', 'p', 'q'.",
+        )
+        self._component = component
+
+        super().__init__(
+            things=coil,
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+            jac_chunk_size=jac_chunk_size,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        from desc.coils import AbstractFiniteBuildCoil
+        from desc.magnetic_fields import SumMagneticField
+
+        coilset = self.things[0]
+
+        finite_build_grid = AbstractFiniteBuildCoil.prep_grid(
+            self._xsection_grid, self._centerline_grid
+        )
+
+        self._dim_f = coilset.num_coils
+        self._constants = {
```

Comment:
I think we are trying to move away from constants #1758 , so would prefer if instead these were all just attributes of the objective self (though I guess `"quad_weights"` is special as we check for it later right in`compute_scaled_error`? so maybe just leave that one but put rest as attributes of self)